### PR TITLE
GraphNode memory leak

### PR DESCRIPTION
**Pull request by gameatp**
Fixes # GraphNode memory leak

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/framework/components/model/component.js [hunk 1]
```diff
@@ -551,12 +557,11 @@ Object.assign(pc, function () {
 
                 var model = new pc.Model();
                 model.graph = node;
-
                 model.meshInstances = [new pc.MeshInstance(node, mesh, this._material)];
 
                 if (system._inTools)
                     model.generateWireframe();
-
+				this._clonedModel = true;
```
Comment (willeastcott) [inline src/framework/components/model/component.js]: Please change tabs for spaces (1 tab === 4 spaces)

File: src/scene/mesh.js [hunk 2]
```diff
@@ -496,7 +496,46 @@ Object.assign(pc, function () {
         syncAabb: function () {
             // Deprecated
         },
+        destroy:function(){
```
Comment (willeastcott) [inline src/scene/mesh.js]: Please try to match the coding style of the engine. For this example: destroy: function () { (notice the spaces)

Comment (willeastcott) [reply src/scene/mesh.js]: Also, leave a black line between the destroy and syncAabb functions.

File: src/scene/mesh.js [hunk 3]
```diff
@@ -496,7 +496,46 @@ Object.assign(pc, function () {
...
+            }
+
+            if (this.morphInstance) {
+                this.morphInstance.destroy();
+                this.morphInstance = null;
+            }
+
+            this.material = null; // make sure instance and material clear references
+            this.node = null;
+        },
```
Comment (willeastcott) [inline src/scene/mesh.js]: Insert blank line.

File: src/framework/components/model/component.js (hunk 1 above)
Comment (daredevildave) [reply src/framework/components/model/component.js]: Why are you setting this? It's not a cloned model?

File: src/framework/components/model/component.js [hunk 4]
```diff
@@ -539,6 +544,7 @@ Object.assign(pc, function () {
                             system.sphere = pc.createSphere(gd, {
                                 radius: 0.5
                             });
+                            system.sphere._refCount++;
```
Comment (daredevildave) [inline src/framework/components/model/component.js]: I'm not sure why this is needed. We shouldn't be accessing private properties in here.

**Comment by daredevildave** [comment]
Can you describe what bug you have fixed here and how? It's not clear as you've moved code around. How is this particle emitter screenshot related to the bug fix?

**Comment by gameatp** [comment]
I did the following tests:
function testShapeModel(){
            var entity = new pc.Entity();
            entity.addComponent("model", {
                type: "sphere"
            });
			entity.setLocalPosition( (Math.random() *2.0 - 1)*5,(Math.random() *2.0 - 1)*5,(Math.random() *2.0 - 1)*5);
            app.root.addChild(entity);
            setTimeout(function(){
				entity.destroy();
            },200);
        }
        for(var n = 0;n < 100;n++)
			testShapeModel();
 
Create 100 spheres and delete them, but the graphNode in reach is not released.
Because the model component has no flag _clonedModel is true

File: src/framework/components/model/component.js (hunk 4 above)
Comment (gameatp) [reply src/framework/components/model/component.js]: If you don't add 1 to the count here, the mesh will actually be released when the model destroy

File: src/scene/mesh.js (hunk 3 above)
Comment (gameatp) [reply src/scene/mesh.js]: ok

File: src/scene/mesh.js (hunk 2 above)
Comment (gameatp) [reply src/scene/mesh.js]: Okay, I'll revise it. I'll pay attention to it next time.

File: src/framework/components/model/component.js (hunk 1 above)
Comment (gameatp) [reply src/framework/components/model/component.js]: If _clonemodel is not true, graphNode will have memory leaks, the model will not call destroy, and the node in meshinstance will not be assigned null.

File: src/scene/mesh.js [hunk 5]
```diff
@@ -497,6 +497,48 @@ Object.assign(pc, function () {
...
+            
+            var mesh,  ib,  j;
+            var device;
+            
+            mesh = this.mesh;
+            if (mesh) {
+                mesh._refCount--;
+                if (mesh._refCount < 1) {
+                    if (mesh.vertexBuffer) {
+                        device = device || mesh.vertexBuffer.device;
```
Comment (willeastcott) [inline src/scene/mesh.js]: ```device``` is not needed. Can be deleted.

File: src/scene/mesh.js [hunk 6]
```diff
@@ -497,6 +497,48 @@ Object.assign(pc, function () {
             // Deprecated
         },
 
+        destroy: function() {
+            
+            var mesh,  ib,  j;
+            var device;
```
Comment (willeastcott) [inline src/scene/mesh.js]: ```device``` is not needed. Can be deleted.

File: src/scene/mesh.js [hunk 7]
```diff
@@ -497,6 +497,48 @@ Object.assign(pc, function () {
...
+            if (mesh) {
+                mesh._refCount--;
+                if (mesh._refCount < 1) {
+                    if (mesh.vertexBuffer) {
+                        device = device || mesh.vertexBuffer.device;
+                        mesh.vertexBuffer.destroy();
+                        mesh.vertexBuffer = null;
+                    }
+                    for (j = 0; j < mesh.indexBuffer.length; j++) {
+                        device = device || mesh.indexBuffer.device;
```
Comment (willeastcott) [inline src/scene/mesh.js]: ```device``` is not needed. Can be deleted.

Comment (gameatp) [reply src/scene/mesh.js]: ok,I repaired it.I found that the set function of mesh may not be released when the mesh reference count is reduced to zero.

**Comment by willeastcott** [comment]
This PR causes the Flappy Bird scene to throw an exception:

https://launch.playcanvas.com/404993

Please run your build of the engine with this scene doing something like:

https://launch.playcanvas.com/404993?use_local_engine=http://localhost/enginegameatp/build/output/playcanvas-latest.js

**Comment by willeastcott** [comment]
@daredevildave I have run through the entire test plan with this PR and it passes.

File: src/framework/components/model/component.js (hunk 4 above)
Comment (willeastcott) [reply src/framework/components/model/component.js]: @daredevildave The only other way prevent a reference counted pc.Mesh from being blown away would be to create a dummy pc.MeshInstance and store that on the model component system. Alternatively, we could create a pc.Model for each primitive and then clone the pc.Model each instantiated primitive-based model component.

Comment (gameatp) [reply src/framework/components/model/component.js]: > I'm not sure why this is needed. We shouldn't be accessing private properties in here.


_refCount is private, and the direct modification here is really not very good, if there is a better way to get the best

Comment (daredevildave) [reply src/framework/components/model/component.js]: > Alternatively, we could create a pc.Model for each primitive (on demand) and then clone the pc.Model for each instantiated primitive-based model component.

This sounds like a better approach to me.

Comment (willeastcott) [reply src/framework/components/model/component.js]: Yeah, I'd agree. It treats primitive-based models like asset-based models.

Comment (gameatp) [reply src/framework/components/model/component.js]: > Yeah, I'd agree. It treats primitive-based models like asset-based models.
image
image
Is there a better way to create meshes in these two places? Is also the way to create a model and then pass clone?

Comment (willeastcott) [reply src/framework/components/model/component.js]: That probably should be how it's done, yes.

**Comment by gameatp** [comment]
> This commit looks like it's fixing a different problem. Maybe submit this in a separate PR?

How to create a new PR before the PR is finished? The only way I submit it is to automatically merge it into the last PR.

**Comment by Maksims** [comment]
When in your repo, you can create branch from remote master branch, something like so:
```
git branch yourNewBranchName upstream/master
git checkout yourNewBranchName
git fetch
git pull
```

It will not include your PR stuff, and will be based on latest master of upstream.

**Pull request closed without merging**